modules/rpc/yarpc.go

```diff
@@ -186,6 +284,38 @@ func RegisterDispatcher(dispatchFn yarpcDispatcherFn) {
 	_dispatcherFn = dispatchFn
 }
 
-func defaultYarpcDispatcher(_ service.Host, cfg yarpc.Config) (*yarpc.Dispatcher, error) {
+func defaultYARPCDispatcher(_ service.Host, cfg yarpc.Config) (*yarpc.Dispatcher, error) {
 	return yarpc.NewDispatcher(cfg), nil
 }
+
+// Dispatcher returns a dispatcher that can be used to create clients.
+// It should be called after all modules have been started, because
+// the each module start creates a new dispatcher and stops previous.
+func Dispatcher() *yarpc.Dispatcher {
+	_dispatcherMu.RLock()
+	defer _dispatcherMu.RUnlock()
+	return _dispatcher
+}
+
+// Stops the current dispatcher, sets the new one with the host, config and _dispatcherFn, and then starts it.
+func resetDispatcher(host service.Host, config yarpc.Config) error {
+	d, err := _dispatcherFn(host, config)
+	if err != nil {
+		return err
+	}
+
+	_dispatcherMu.Lock()
+	defer _dispatcherMu.Unlock()
+
+	if _dispatcher == nil {
+		_dispatcher = d
+		return _dispatcher.Start()
+	}
+
+	if err := _dispatcher.Stop(); err != nil {
```

Comment:
Is the goal to be able to start/stop fx modules individually during the lifetime of the application?  Or is this for the shutdown/drain of the entire application?

If this is for the shutdown of the entire application, in YARPC we can update dispatcher.Stop/Start() to be idempotent so we can shutdown all modules at the same time, and they will all block on a sync.Once until the dispatcher shuts down.  (We could also wrap the dispatcher start/stop in uberfx to do the same)

We currently do this for outbounds/inbounds in YARPC using our LifecycleOnce abstraction, it can be added to the dispatcher as well.
https://github.com/yarpc/yarpc-go/blob/dev/internal/sync/lifecycle.go#L58

If this is for starting/stopping during the lifetime of the application, what is the use case we want to support?

The current implementation here will probably lead to some unknown/strange behavior since we'll be starting/stopping the inbounds dynamically which is not supported by YARPC inbounds/outbounds.